Zvk vector crypto support (v5)

egouriou-rivos opened the pull request:
[Updated 2023-06-02]

This patch series implements Zvk at version Version v0.9.7, 31 May, 2023, from <https://github.com/riscv/riscv-crypto/tree/master/doc/vector>:
 - Zvbb
 - Zvbc
 - Zvkg
 - Zvkned
 - Zvknha / Zvknhb
 - Zvksed
 - Zvksh
no-instruction extension (only accepted as part of the ISA string, no-op otherwise):
 - Zvkt
and combo extensions (expanded to the underlying extensions):
 - Zvkn
 - Zvknc
 - Zvkng
 - Zvks
 - Zvksc
 - Zvksg

Some Zvk instructions conflict with instructions part of Zvpn, part of the proposed packed SIMD extensions. The conflicting instructions are listed in the overlapped file and only enabled when their extension is declared. Extension in conflicts are detected in isa_parser.cc, triggering errors.

The implementation of element groups does not support WORDS_BIGENDIAN correctly. An error is emitted when WORDS_BIGENDIAN is enabled together with extensions using element group accesses.

egouriou-rivos commented:
@aswaterman Andrew, this is my first PR in this repository. As such I am quite unfamiliar with the review process. Don't hesitate to point me in the right direction. Thanks.

scottj97 commented on riscv/encoding.h:
```diff
@@ -4,7 +4,7 @@
 /*
  * This file is auto-generated by running 'make' in
- * https://github.com/riscv/riscv-opcodes (02b4866)
+ * https://github.com/riscv/riscv-opcodes (b6fea6c)
```
Please update the URL here if this was not generated from the official riscv-opcodes repo.

egouriou-rivos replied:
Done, thanks for noticing. It's a little strange to edit an auto-generated file, but it makes sense for now.

scottj97 commented on riscv/encoding.h:
```diff
@@ -381,8 +381,12 @@
 #define MASK_ADD8 0xfe00707f
 #define MATCH_ADD_UW 0x800003b
 #define MASK_ADD_UW 0xfe00707f
+#define MATCH_ADDD 0x7b
```
5f2a722f272ccd4ef69ec1c8c12edd1ef1648948 mentions removal of 2 P-extension instructions (pkbb32, pktt32) but I don't see that in the diff?

egouriou-rivos replied:
The P extension evolved between the time I created the commit and the time I sent the PR. That comment was obsolete, thanks for calling it out.

scottj97 commented on riscv/riscv.mk.in:
```diff
@@ -1329,7 +1329,6 @@ riscv_insn_list = \
 	$(if $(HAVE_INT128),$(riscv_insn_ext_v),) \
 	$(riscv_insn_ext_zce) \
 	$(riscv_insn_ext_h) \
-	$(riscv_insn_ext_p) \
```
This commit also needs to disallow P in `isa_parser.cc`, so that nobody is confused about whether P is supported or not.

scottj97 commented on riscv/zvk_ext_macros.h:
```diff
@@ -0,0 +1,1251 @@
+// Ensures that the ZVBB extension (vector crypto bitmanip) is present,
+// which depends on vector.
+#define require_zvbb \
+  do { \
+    require_vector(true); \
+    require_extension(EXT_ZVBB); \
+  } while (0)
```
If Zvbb requires V, that should be enforced in `isa_parser.cc`. It does not need to be enforced here in every instruction. (Same comment for all of this family of macros; it looks like none of these macros are needed.)

nibrunieAtSi5 replied:
I think the point here is to raise an exception when an instruction is executed without proper support. I am not that familiar with spike, but I would imagine `require_extension(EXT_ZVBB);` is required to raise the exception if the instruction is not supported. The dependency on V (which may be V or Zve* I think) should indeed be handled when the ISA support is initialized.

aswaterman replied:
The `require_vector(true)` checks `sstatus.VS` and `vtype.vill`.  So, even though the ISA parser should enforce the extension dependency, this code (or something similar) is still necessary.

scottj97 replied:
> The `require_vector(true)` checks `sstatus.VS` and `vtype.vill`. So, even though the ISA parser should enforce the extension dependency, this code (or something similar) is still necessary.

Good point. Should those checks be part of `require_extension(EXT_ZVBB)`, do you think?

aswaterman replied:
IMO, the current formulation is OK.

The only problem with it, I think, is that it assumes Zvbb requires V.  This isn't strictly true; Zvbb only requires a subset of V: Zve64x, IIRC.  But since Spike doesn't support any of the V subsets currently, that is not a problem currently.

Furthermore, if Spike is augmented to support Zve64x etc., then essentially all of the vector instructions will need to be retooled in the same manner as this one.  In other words, this approach doesn't create a new type of technical debt.

egouriou-rivos replied:
Thanks for the discussion. I updated the documentation of those new require_xyz macros to state 
```
// Ensures that the ... extension is present,
// and the vector unit is enabled and in a valid state.```

This captures the semantics of require_vector(true) more clearly than the previous "and that vector is present".

scottj97 commented on riscv/zvk_ext_macros.h:
```diff
@@ -0,0 +1,1251 @@
+    require_extension(EXT_ZVKSH); \
+  } while (0)
+
+// Ensures that the vector instruction is not using a mask.
+#define require_no_vmask  require(insn.v_vm() == 1)
+
+// Ensures that the current VSEW is the given N number of bits.
+#define require_vsew(N)  require(P.VU.vsew == (N))
```
These two are so trivial and self-expanatory that I'd say they should just be inlined.

egouriou-rivos replied:
I agree for `require_vsew()` and got rid of them.

I have a stronger attachment to `require_no_vmask` as I find it a lot clearer than `require(insn.v_vm() == 1)`, especially since VM is a highly overloaded acronym and "VM == 1" meaning "vector mask is not used" is very counter-intuitive.  
It is a case where someone intimate with the vector spec may find the "insn.v_vm() == 1" very clear but newcomers or more casual contributors will either misinterpret or need to go back to the spec to understand. When I wrote that code I very much needed the more explicit semantics.

egouriou-rivos replied:
Let me know if you still want me to remove the `require_no_vmask`.

scottj97 commented on riscv/zvk_ext_macros.h:
```diff
@@ -0,0 +1,1251 @@
+// containing element groups of 8 uint32_t values (EGW=256, EEW=32, EGS=8),
+// *ignoring* the current SEW that applies to the vectors.
+//
+// IMPORTANT
```
Why is this info important? Isn't it obvious from reading the code?

scottj97 commented on riscv/zvk_ext_macros.h:
```diff
@@ -0,0 +1,1251 @@
+      switch (sew) { \
+        case e8: { \
+          VI_ZVK_VI_WIDENING_U_PARAMS(e8); \
+          do { BODY } while (0); \
```
Any reason this line can't simply be:
```
BODY; \
```

egouriou-rivos replied:
No good reason, this came from macro paranoia. After more consideration this `do { } while (0)` is not achieving anything. I removed them and updated the macro documentation to reflect the change.

Thanks for calling this out.

scottj97 commented on riscv/zvk_ext_macros.h:
```diff
@@ -0,0 +1,1251 @@
+// into the (mutable) variables named by the W arguments, provided in
+// "Little Endian" (LE) order, i.e., from the least significant (W0)
+// to the most significant (W3).
+#define EXTRACT_EGU32x4_WORDS_LE(X, W0, W1, W2, W3) \
```
Many of the macros seem like they would be better off as methods on a `class EGU32x4_t`.

egouriou-rivos replied:
I'm 100% with you on the desire to reduce macro usage. I wish I had known that when I started writing that code, instead of following the vector idioms.

EGU32x4_t ends up being std:array<uint32_t, 4>, and using instance methods is not going to work as-is. We can use freestanding methods. We could hide the std::array<> as an implementation detail of a class but it will require forwarding quite a few operations.

scottj97 reviewed (changes requested):
I didn't look at the details of the instructions, but generally I'd say this uses too many macros when compilable C++ code would suffice. (The existing instructions also overuse macros, IMO; let's try to improve that, not make it worse.)

@aswaterman is there a good example of another extension that uses custom data types and/or helper functions instead of macros?

aswaterman commented:
@scottj97 As a very small example, adding the various `xperm` instructions was accomplished by adding a new function to `arith.h`: https://github.com/riscv-software-src/riscv-isa-sim/pull/629/files

I fully agree that, in cases where we can use functions, possibly with templates, rather than macros, we should.  Although Spike is full of examples to the contrary, we are trying to move away from that style.

aswaterman commented:
As for handling overlapping opcodes, the problematic ones need to be added to the overlap list; here's an example: https://github.com/riscv-software-src/riscv-isa-sim/commit/087626c280571b13e5ee51a3a7603c8bbd96916a#diff-19e2a333b37ddb8e4a5ec7706ce201da3561f4ff223756fa8b9e692ea4679a8fR10

nibrunieAtSi5 commented on riscv/zvkned_ext_macros.h:
```diff
@@ -0,0 +1,270 @@
+// vaes*.vv instruction constraints. Those are the same as the .vs ones,
+// except for the overlap constraint that is not present for .vv variants.
+//  - Zvkned is enabled
+//  - LMUL * VLEN <= EGW (128)
```
Shouldn't this be
```suggestion
//  - EGW (128) <= LMUL * VLEN 
```

nibrunieAtSi5 commented on riscv/insns/vaesef_vv.h:
```diff
@@ -0,0 +1,61 @@
+#include "zvkned_ext_macros.h"
+#include "zvk_ext_macros.h"
+
+// Uncomment to enble debug logging of invocations of this instruction.
```
```suggestion
// Uncomment to enable debug logging of invocations of this instruction.
```

egouriou-rivos replied:
Done here and for all the cut/pasted instances. Thanks for catching this.

nibrunieAtSi5 commented on riscv/insns/vaesef_vs.h:
```diff
@@ -0,0 +1,66 @@
+#include "zvkned_ext_macros.h"
+#include "zvk_ext_macros.h"
+
+// Uncomment to enble debug logging of invocations of this instruction.
```
```suggestion
// Uncomment to enable debug logging of invocations of this instruction.
```

nibrunieAtSi5 commented on riscv/insns/vaesdm_vv.h:
```diff
@@ -0,0 +1,63 @@
+#include "zvkned_ext_macros.h"
+#include "zvk_ext_macros.h"
+
+// Uncomment to enble debug logging of invocations of this instruction.
```
```suggestion
// Uncomment to enable debug logging of invocations of this instruction.
```

nibrunieAtSi5 commented on riscv/insns/vaesdm_vs.h:
```diff
@@ -0,0 +1,69 @@
+#include "zvkned_ext_macros.h"
+#include "zvk_ext_macros.h"
+
+// Uncomment to enble debug logging of invocations of this instruction.
```
```suggestion
// Uncomment to enable debug logging of invocations of this instruction.
```

nibrunieAtSi5 commented on riscv/insns/vwsll_vi.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "zvk_ext_macros.h"
+
+// Uncomment to enble debug logging of invocations of this instruction.
```
```suggestion
// Uncomment to enable debug logging of invocations of this instruction.
```

nibrunieAtSi5 commented on riscv/insns/vwsll_vv.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "zvk_ext_macros.h"
+
+// Uncomment to enble debug logging of invocations of this instruction.
```
```suggestion
// Uncomment to enable debug logging of invocations of this instruction.
```

nibrunieAtSi5 commented on riscv/insns/vwsll_vx.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "zvk_ext_macros.h"
+
+// Uncomment to enble debug logging of invocations of this instruction.
```
```suggestion
// Uncomment to enable debug logging of invocations of this instruction.
```

scottj97 commented:
@egouriou-rivos can you tell us how you've tested this?

charmitro commented:
Just a heads-up for anyone following this, as this is needed for [ACT testing](https://github.com/riscv-non-isa/riscv-arch-test/pull/333).

The SAIL implementation is written for [Vector Crypto Draft v20230303](https://github.com/riscv/riscv-crypto/releases/download/v20230303/riscv-crypto-spec-vector.pdf) which leads to incorrect ACT testing.

In order to have some(not all Extensions are possible) valid results between DUT(Spike) & Sail Signatures you'd have to apply a fair amount of workarounds so the ACT tooling can match the correct Extensions between **Spike <-> SAIL**.

Successful Test reports could be generated only for the following extensions:
- Zvkg
- Zvksed
- Zvksh
- Zvkned

Failing Test reports, with the reason, are:
- Zvkb - Naming conventions and instructions included in extension changed in newer version
- Zvknhab - Not able to run at all with the ACT tooling. Even executing `spike $(which pk) --varch=vlen:256,elen:32 --isa=rv64imfdcvzvknha_zvknhb a.out` doesn't seem to work. LMK if this is a user error on my side.

egouriou-rivos commented:
> @scottj97 As a very small example, adding the various `xperm` instructions was accomplished by adding a new function to `arith.h`: https://github.com/riscv-software-src/riscv-isa-sim/pull/629/files
> 
> I fully agree that, in cases where we can use functions, possibly with templates, rather than macros, we should. Although Spike is full of examples to the contrary, we are trying to move away from that style.

I am overjoyed to read this, as getting into the vector code for this extension was quite frustrating due to those layers of (unhygienic) macros and the utter lack of documentation of their intent. Unfortunately the Zvk implementation strove to fit within the existing idioms we found, with the exception of (over-)documenting.

I will not have the time to perform the suggested cleanup until later this year. I feel that such a cleanup needs to involve the V instructions and will represent a significant effort. As such, I would like to see it decoupled from landing the Zvk support.

egouriou-rivos commented:
> Also please fix the CI failure:
> 
> ```
> /home/runner/work/riscv-isa-sim/riscv-isa-sim/ci-tests/../riscv/insns/sm4_common.h:29:23: error: unused variable ‘ck’ [-Werror=unused-variable]
>    29 | static const uint32_t ck[32] = {
>       |                       ^~
> ```

This should be fixed now.

egouriou-rivos commented:
I pushed updated commits with the following changes:
 - Fixed vwsll's shift amount mask to use 2*SEW bits (vs SEW previously), matching a spec pseudo-code fix.
 - Fixed repeated typos mentioned above
 - Deleted the `require_vsew` macro, per Scott's suggestion
 - Moved the SM4 `ck` array from sm4_common.h to the only use location, `vsm4k.vi`, resolving a compiler warning in CI
 - Manually fixed the URL in encoding.h to point to the github.com/rivosinc/riscv-opcode source 
 - Removed useless "do { } while (0)" wrapping the BODY blocks, per Scott's review

With this, the PR is up to date with the current specification, which should allow for the spec to move into freeze for this deliverable.

Issues I intend to address in upcoming versions:
 - Properly take care of the P / Zvk conflicts, per Andrew's suggestion above
 - Bring the encodings from the official riscv/riscv-opcodes repo (the PR just landed, I'll fix the reference in the next push, https://github.com/riscv/riscv-opcodes/pull/154)

I would like to defer the cleanup from C-style macros to more idiomatic C++ to subsequent PRs. As discussed above, this PR was built over many months following the example of the V extension which was considered idiomatic. I wish I had checked when that work started.

egouriou-rivos commented:
One issue I am aware of and would like recommendation on: the element-group accessor `vectorUnit_t::elt_grp<>(...)` is currently broken when WORDS_BIGENDIAN is defined. I have a hard time mapping big-endian to the element groups and the vector register layout. 

`vectorUnit_t::elt<>(...)` contains the following logic to account for BIGENDIAN:
```
#ifdef WORDS_BIGENDIAN
  // "V" spec 0.7.1 requires lower indices to map to lower significant
  // bits when changing SEW, thus we need to index from the end on BE.
  n ^= elts_per_reg - 1;
#endif
```

I worry that no such trick will allow us to fix `elt_grp` to work with BIGENDIAN. I may need to access the elements separately to reproduce a similar effect instead of performing a contiguous mapping to a std::array.

egouriou-rivos commented:
I uploaded a new version of this series. Differences from v2:
 - Properly (?) dealt with conflicts between P (Zpn) and Zvk instructions, removing the need to disable P in riscv.mk.in,
 - Imported encodings from the official riscv-opcodes repository where those landed in master/ (https://github.com/riscv/riscv-opcodes/commit/199e646f2a8e36),
 - Added handling of combo extensions Zvkng/Zvksg and no-op Zvkt,
 - Fixed one pre-existing typo in isa_parser.cc

As mentioned above, the one known issue is the handling of WORDS_BIGENDIAN. I'd appreciate if people familiar with big-endian and the vector unit register layout could give some thought as to whether elt_group<>(...) is implementable for big-endian or if we need to change that interface.

egouriou-rivos commented:
I uploaded a new version of this series (v4). The only difference from the previous version  is in 'isa_parsing.cc' in  the commit titled 'Zvk extension parsing':
 - Removed Zvbc from the combo extensions Zvkn, Zvks
 - By transitive closure, removed Zvbc from the combo extensions Zvkng, Zvksg

This matches the changes that went into the specification in version 0.9.5, 2023-05-09
(commit https://github.com/riscv/riscv-crypto/commit/49a2d3eb95552254906f0a783ddb0f683de1fd54, release https://github.com/riscv/riscv-crypto/releases/tag/v20230509)

aswaterman commented on riscv/encoding.h:
```diff
@@ -4,7 +4,7 @@
 /*
  * This file is auto-generated by running 'make' in
- * https://github.com/riscv/riscv-opcodes (02b4866)
+ * https://github.com/riscv/riscv-opcodes (7876224)
```
Since `encoding.h` has been updated after you created this PR, can you rebase on top of the HEAD of master and make sure the history is clean?

egouriou-rivos replied:
Done. 7876224 is still more recent than what is currently in riscv-isa-sim's master.

aswaterman commented on riscv/riscv.mk.in:
```diff
@@ -1315,26 +1315,29 @@ riscv_insn_ext_zicond = \
 
+# Note that riscv_insn_ext_p and riscv_insn_ext_zvk contain instructions
+# that have conflicting encodings. They cannot be both included concurrently.
+# TODO: have a mechanism to enable those groups at runtime.
```
I see you added this comment here and removed it in the next PR.  Can you instead delete the comment here so it doesn't appear in the history?

egouriou-rivos replied:
Done, thanks for catching this.

aswaterman commented on riscv/vector_unit.cc:
```diff
@@ -86,6 +86,54 @@ template<class T> T& vectorUnit_t::elt(reg_t vReg, reg_t n, bool UNUSED is_write
+vectorUnit_t::elt_group(reg_t vReg, reg_t n, bool UNUSED is_write) {
+#ifdef WORDS_BIGENDIAN
+# error This logic is invalid for Big Endian mode.
+#endif
```
This is problematic since we have big-endian users.  Two options:
- The obvious one, make this logic correct for big-endian hosts.
- Keep the old big-endian functionality working, and disallow use of this functionality on big-endian hosts.  You can have something like

```
#ifdef WORDS_BIGENDIAN
  abort();
#endif
```

and then also change the `isa_parser` to disallow any extensions that rely on this when `WORDS_BIGENDIAN` is defined.

egouriou-rivos replied:
Done, thanks for the suggestion.

aswaterman commented on riscv/insns/vrol_vx.h:
```diff
@@ -0,0 +1,21 @@
+
+VI_VX_ULOOP
+({
+  if (shift != 0) {  // Protect against UB.
+    vd = (vs2 << shift) | (vs2 >> (sew - shift));
+  } else {
+    vd = vs2;
+  }
```
I think this would be cleaner as `vd = (vs2 << shift) | (vs2 >> (-shift & (sew - 1)));`, which fixes the UB in the same way as we did for the scalar `rol` instruction.

egouriou-rivos replied:
Done, thanks for the suggestion.

I did a bit more work around the rotate logic in those commits:
 - I introduced rotate_right/rotate_left template functions in arith.h . 
 - I use those functions in the ROR/ROL macros in zvk_ext_macros.h .
 - I added "ZVK_" prefixes to the ROR/ROL macros to reduce the risk of collisions

The vector instructions still use their own logic rather than use the rotate_right/rotate_left versions. Let me know if you think we should change that.

aswaterman commented on riscv/insns/vror_vi.h:
```diff
@@ -0,0 +1,21 @@
+
+VI_VX_ULOOP
+({
+  if (shift != 0) {  // Protect against UB.
+    vd = (vs2 >> shift) | (vs2 << (sew - shift));
+  } else {
+    vd = vs2;
+  }
```
Same comment as https://github.com/riscv-software-src/riscv-isa-sim/pull/1303/commits/68bb2dbb3c29b111c5b7ef96f0c32b544bfc22c6#r1199403904

egouriou-rivos replied:
Done.

aswaterman commented on riscv/insns/vrol_vv.h:
```diff
@@ -0,0 +1,20 @@
+    vd = (vs2 << shift) | (vs2 >> (sew - shift));
+  } else {
+    vd = vs2;
+  }
```
Same comment as https://github.com/riscv-software-src/riscv-isa-sim/pull/1303/commits/68bb2dbb3c29b111c5b7ef96f0c32b544bfc22c6#r1199403904

egouriou-rivos replied:
Done

aswaterman commented on riscv/insns/vror_vx.h:
```diff
@@ -0,0 +1,21 @@
+    vd = (vs2 >> shift) | (vs2 << (sew - shift));
+  } else {
+    vd = vs2;
+  }
```
Same comment as https://github.com/riscv-software-src/riscv-isa-sim/pull/1303/commits/68bb2dbb3c29b111c5b7ef96f0c32b544bfc22c6#r1199403904

egouriou-rivos replied:
Done

aswaterman commented on riscv/insns/vror_vv.h:
```diff
@@ -0,0 +1,20 @@
+    vd = (vs2 >> shift) | (vs2 << (sew - shift));
+  } else {
+    vd = vs2;
+  }
```
Same comment as https://github.com/riscv-software-src/riscv-isa-sim/pull/1303/commits/68bb2dbb3c29b111c5b7ef96f0c32b544bfc22c6#r1199403904

egouriou-rivos replied:
Done.

aswaterman commented on riscv/insns/vwsll_vi.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "zvk_ext_macros.h"
+
+// Uncomment to enable debug logging of invocations of this instruction.
+// #define DLOG_INVOCATION
+
+#if defined(DLOG_INVOCATION)
+#define DLOG(...) ZVK_DBG_LOG(__VA_ARGS__)
+// Print format/value for both state and round key element groups.
+#define PRIxEG PRIxEGU8x16
+#define PRVEG(X) PRVEGU8x16_LE(X)
+#else
+#define DLOG(...) (void)(0)
+#endif
+
+require_zvbb;
+
+DLOG("vwssl_vi vd#(%" PRIuREG "), vs2#(%" PRIuREG "), zimm5(%" PRIuREG ")"
+     " vstart(%" PRIuREG ") vl(%" PRIuREG ")",
+     insn.rd(), insn.rs2(), insn.v_zimm5(),
+     P.VU.vstart->read(), P.VU.vl->read());
+
```
Delete.

egouriou-rivos replied:
Done. I removed the debug logging throughout.

I understand that this logic was noisy and the desire to remove it. Please note however that developing this logging was critical to implement this functionality. The logs were modeled to closely match format used by each cryptographic algorithm specification (AES, SHA-2, SM3, SM4) in the respective sample sections. The logs found quite a few issues with the Zvk specification. As the Zvk specification is stable, its usefulness may indeed be over. Good luck to anyone trying to make significant changes to those instruction implementations in the absence of those logs.

aswaterman commented on riscv/insns/vwsll_vi.h:
```diff
@@ -0,0 +1,27 @@
+#define PRVEG(X) PRVEGU8x16_LE(X)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vwsll_vv.h:
```diff
@@ -0,0 +1,27 @@
+#define PRVEG(X) PRVEGU8x16_LE(X)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vwsll_vv.h:
```diff
@@ -0,0 +1,27 @@
+DLOG("vwssl_vv vd#(%" PRIuREG "), vs2#(%" PRIuREG "), vs1#(%" PRIuREG ")"
+     " vstart(%" PRIuREG ") vl(%" PRIuREG ")",
+     insn.rd(), insn.rs2(), insn.rs1(),
+     P.VU.vstart->read(), P.VU.vl->read());
```
Delete.

egouriou-rivos replied:
Done. I'll just mark all similar comments/suggestions/requests as resolved.

aswaterman commented on riscv/insns/vwsll_vx.h:
```diff
@@ -0,0 +1,27 @@
+#define PRVEG(X) PRVEGU8x16_LE(X)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vwsll_vx.h:
```diff
@@ -0,0 +1,27 @@
+DLOG("vwssl_vx vd#(%" PRIuREG "), vs2#(%" PRIuREG "), rs1#(%" PRIuREG ")"
+     " vstart(%" PRIuREG ") vl(%" PRIuREG ")",
+     insn.rd(), insn.rs2(), insn.rs1(),
+     P.VU.vstart->read(), P.VU.vl->read());
```
Delete.

aswaterman commented on riscv/insns/vclmul_vv.h:
```diff
@@ -0,0 +1,31 @@
+#define DLOG(...) ZVK_DBG_LOG(__VA_ARGS__)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vclmul_vv.h:
```diff
@@ -0,0 +1,31 @@
+    }
+  }
+  DLOG("vclmul_vv vd(%" PRIx64 ") <- vs2(%" PRIx64 ") x vs1(%" PRIx64 ")",
+       (uint64_t)vd, (uint64_t)vs2, (uint64_t)vs1);
```
Delete.

aswaterman commented on riscv/insns/vghsh_vv.h:
```diff
@@ -0,0 +1,54 @@
+#define PRV_R_EG(reg_num, reg) PRV_REG_EGU32x4_LE(reg_num, reg)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vghsh_vv.h:
```diff
@@ -0,0 +1,54 @@
+  {
+    DLOG("vghsh Y=" PRI_uR_xEG " H=" PRI_uR_xEG " X=" PRI_uR_xEG,
+         PRV_R_EG(vd_num, vd), PRV_R_EG(vs2_num, vs2),
+         PRV_R_EG(vs1_num, vs1));
```
Delete.

aswaterman commented on riscv/insns/vgmul_vv.h:
```diff
@@ -0,0 +1,47 @@
+#define PRV_R_EG(reg_num, reg) PRV_REG_EGU32x4_LE(reg_num, reg)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vgmul_vv.h:
```diff
@@ -0,0 +1,47 @@
+  {},
+  {
+    DLOG("vgmul Y=" PRI_uR_xEG " H=" PRI_uR_xEG,
+         PRV_R_EG(vd_num, vd), PRV_R_EG(vs2_num, vs2));
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+#define DLOG(...) ZVK_DBG_LOG(__VA_ARGS__)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+// Print format/value for "v<reg_num>(<Element Group in Hex, Big Endian>)"
+#define PRI_uR_xEG PRI_uREG_xEGU32x4
+#define PRV_R_EG(reg_num, reg) PRV_REG_EGU32x4_BE(reg_num, reg)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+    VI_ZVK_VD_VS1_VS2_EGU32x4_NOVM_LOOP(
+      {
+         DLOG("-- vsha2ch_vv(256) " ZVK_PRI_REGNUMS_VD_VS2_VS1,
+              ZVK_PRV_REGNUMS_VD_VS2_VS1);
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+      {
+        DLOG("vsha2ch " PRI_uR_xEG " " PRI_uR_xEG " " PRI_uR_xEG,
+             PRV_R_EG(vd_num, vd), PRV_R_EG(vs2_num, vs2),
+             PRV_R_EG(vs1_num, vs1));
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+
+        // Update the destination register, vd <- {a, b, e, f}.
+        SET_EGU32x4_BE(vd, a, b, e, f);
+        DLOG("= vsha2ch " PRI_uR_xEG, PRV_R_EG(vd_num, vd));
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+// Print format/value for "v<reg_num>(<Element Group in Hex, Big Endian>)"
+#define PRI_uR_xEG PRI_uREG_xEGU64x4
+#define PRV_R_EG(reg_num, reg) PRV_REG_EGU64x4_BE(reg_num, reg)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+    );
+    break;
+#undef PRI_uR_xEG
+#undef PRV_R_EG
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+    VI_ZVK_VD_VS1_VS2_EGU64x4_NOVM_LOOP(
+      {
+        DLOG("-- vsha2ch_vv(512) " ZVK_PRI_REGNUMS_VD_VS2_VS1,
+             ZVK_PRV_REGNUMS_VD_VS2_VS1);
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+      {
+        DLOG("vsha2ch " PRI_uR_xEG " " PRI_uR_xEG " " PRI_uR_xEG,
+             PRV_R_EG(vd_num, vd), PRV_R_EG(vs2_num, vs2),
+             PRV_R_EG(vs1_num, vs1));
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+
+        // Update the destination register, vd <- {a, b, e, f}.
+        SET_EGU64x4_BE(vd, a, b, e, f);
+        DLOG("= vsha2ch " PRI_uR_xEG, PRV_R_EG(vd_num, vd));
```
Delete.

aswaterman commented on riscv/insns/vsha2ch_vv.h:
```diff
@@ -0,0 +1,101 @@
+    );
+    break;
+#undef PRI_uR_xEG
+#undef PRV_R_EG
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+#define DLOG(...) ZVK_DBG_LOG(__VA_ARGS__)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+// Print format/value for "v<reg_num>(<Element Group in Hex, Big Endian>)"
+#define PRI_uR_xEG PRI_uREG_xEGU32x4
+#define PRV_R_EG(reg_num, reg) PRV_REG_EGU32x4_BE(reg_num, reg)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+    VI_ZVK_VD_VS1_VS2_EGU32x4_NOVM_LOOP(
+      {
+         DLOG("-- vsha2cl_vv(256) " ZVK_PRI_REGNUMS_VD_VS2_VS1,
+              ZVK_PRV_REGNUMS_VD_VS2_VS1);
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+      {
+        DLOG("vsha2cl " PRI_uR_xEG " " PRI_uR_xEG " " PRI_uR_xEG,
+             PRV_R_EG(vd_num, vd), PRV_R_EG(vs2_num, vs2),
+             PRV_R_EG(vs1_num, vs1));
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+
+        // Update the destination register, vd <- {a, b, e, f}.
+        SET_EGU32x4_BE(vd, a, b, e, f);
+        DLOG("= vsha2cl " PRI_uR_xEG, PRV_R_EG(vd_num, vd));
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+    );
+    break;
+#undef PRI_uR_xEG
+#undef PRV_R_EG
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+// Print format/value for "v<reg_num>(<Element Group in Hex, Big Endian>)"
+#define PRI_uR_xEG PRI_uREG_xEGU64x4
+#define PRV_R_EG(reg_num, reg) PRV_REG_EGU64x4_BE(reg_num, reg)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+    VI_ZVK_VD_VS1_VS2_EGU64x4_NOVM_LOOP(
+      {
+        DLOG("-- vsha2cl_vv(512) " ZVK_PRI_REGNUMS_VD_VS2_VS1,
+             ZVK_PRV_REGNUMS_VD_VS2_VS1);
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+      {
+        DLOG("vsha2cl " PRI_uR_xEG " " PRI_uR_xEG " " PRI_uR_xEG,
+             PRV_R_EG(vd_num, vd), PRV_R_EG(vs2_num, vs2),
+             PRV_R_EG(vs1_num, vs1));
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+
+        // Update the destination register, vd <- {a, b, e, f}.
+        SET_EGU64x4_BE(vd, a, b, e, f);
+        DLOG("= vsha2cl " PRI_uR_xEG, PRV_R_EG(vd_num, vd));
```
Delete.

aswaterman commented on riscv/insns/vsha2cl_vv.h:
```diff
@@ -0,0 +1,101 @@
+    );
+    break;
+#undef PRI_uR_xEG
+#undef PRV_R_EG
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+#define DLOG(...) ZVK_DBG_LOG(__VA_ARGS__)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+// Print format/value for "v<reg_num>(<Element Group in Hex, Big Endian>)"
+#define PRI_uR_xEG PRI_uREG_xEGU32x4
+#define PRV_R_EG(reg_num, reg) PRV_REG_EGU32x4_BE(reg_num, reg)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+    VI_ZVK_VD_VS1_VS2_EGU32x4_NOVM_LOOP(
+      {
+         DLOG("-- vsha2ms_vv(256) " ZVK_PRI_REGNUMS_VD_VS2_VS1,
+              ZVK_PRV_REGNUMS_VD_VS2_VS1);
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+      {
+        DLOG("vsha2ms " PRI_uR_xEG " " PRI_uR_xEG " " PRI_uR_xEG,
+             PRV_R_EG(vd_num, vd), PRV_R_EG(vs2_num, vs2),
+             PRV_R_EG(vs1_num, vs1));
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+
+        // Update the destination register.
+        SET_EGU32x4_BE(vd, w19, w18, w17, w16);;
+        DLOG("= vsha2ms " PRI_uR_xEG, PRV_R_EG(vd_num, vd));
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+    );
+    break;
+#undef PRI_uR_xEG
+#undef PRV_R_EG
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+// Print format/value for "v<reg_num>(<Element Group in Hex, Big Endian>)"
+#define PRI_uR_xEG PRI_uREG_xEGU64x4
+#define PRV_R_EG(reg_num, reg) PRV_REG_EGU64x4_BE(reg_num, reg)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+    VI_ZVK_VD_VS1_VS2_EGU64x4_NOVM_LOOP(
+      {
+        DLOG("-- vsha2ms_vv(512) " ZVK_PRI_REGNUMS_VD_VS2_VS1,
+             ZVK_PRV_REGNUMS_VD_VS2_VS1);
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+      {
+        DLOG("vsha2ms " PRI_uR_xEG " " PRI_uR_xEG " " PRI_uR_xEG,
+             PRV_R_EG(vd_num, vd), PRV_R_EG(vs2_num, vs2),
+             PRV_R_EG(vs1_num, vs1));
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+
+        // Update the destination register.
+        SET_EGU64x4_BE(vd, w19, w18, w17, w16);;
+        DLOG("= vsha2ms " PRI_uR_xEG, PRV_R_EG(vd_num, vd));
```
Delete.

aswaterman commented on riscv/insns/vsha2ms_vv.h:
```diff
@@ -0,0 +1,104 @@
+    );
+    break;
+#undef PRI_uR_xEG
+#undef PRV_R_EG
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vs.h:
```diff
@@ -0,0 +1,67 @@
+#define PRVEG(X) PRVEGU8x16_LE(X)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vs.h:
```diff
@@ -0,0 +1,67 @@
+  {
+    DLOG("vaesef_vs vd#(%" PRIuREG ") vs2#(%" PRIuREG ")"
+         " vstart_eg(%" PRIuREG ") vl_eg(%" PRIuREG ")",
+         vd_num, vs2_num, vstart_eg, vl_eg);
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vs.h:
```diff
@@ -0,0 +1,67 @@
+
+    DLOG("vaesdf: @start state(%" PRIxEG "), scalar_key(%" PRIxEG ")"
+         " idx_eg(%" PRIuREG ")",
+         PRVEG(aes_state), PRVEG(scalar_key), idx_eg);
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vs.h:
```diff
@@ -0,0 +1,67 @@
+    VAES_INV_SUB_BYTES(aes_state);
+    DLOG("vaesdf: @is_box state(%" PRIxEG ")", PRVEG(aes_state));
+    // AddRoundKey (which is also InvAddRoundKey as it's xor)
+    DLOG("vaesdf: ik_sched round_key(%" PRIxEG ")", PRVEG(scalar_key));
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vs.h:
```diff
@@ -0,0 +1,67 @@
+
+    // InvShiftRows - Rotate each row bytes by 0, 1, 2, 3 positions.
+    VAES_INV_SHIFT_ROWS(aes_state);
+    DLOG("vaesdf: @is_row state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vs.h:
```diff
@@ -0,0 +1,67 @@
+    DLOG("vaesdf: @is_row state(%" PRIxEG ")", PRVEG(aes_state));
+    // InvSubBytes - Apply S-box to every byte in the state
+    VAES_INV_SUB_BYTES(aes_state);
+    DLOG("vaesdf: @is_box state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vs.h:
```diff
@@ -0,0 +1,67 @@
+    // AddRoundKey (which is also InvAddRoundKey as it's xor)
+    DLOG("vaesdf: ik_sched round_key(%" PRIxEG ")", PRVEG(scalar_key));
+    EGU8x16_XOREQ(aes_state, scalar_key);
+    DLOG("vaesdf: @ik_add/@ioutput state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vv.h:
```diff
@@ -0,0 +1,61 @@
+#define PRVEG(X) PRVEGU8x16_LE(X)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vv.h:
```diff
@@ -0,0 +1,61 @@
+  {
+    DLOG("vaesdf_vv vd#(%" PRIuREG ") vs2#(%" PRIuREG ")"
+         " vstart_eg(%" PRIuREG ") vl_eg(%" PRIuREG ")",
+         vd_num, vs2_num, vstart_eg, vl_eg);
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vv.h:
```diff
@@ -0,0 +1,61 @@
+
+    DLOG("vaesdf: @start state(%" PRIxEG "), key(%" PRIxEG ")"
+         " idx_eg(%" PRIuREG ")",
+         PRVEG(aes_state), PRVEG(round_key), idx_eg);
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vv.h:
```diff
@@ -0,0 +1,61 @@
+
+    // InvShiftRows - Rotate each row bytes by 0, 1, 2, 3 positions.
+    VAES_INV_SHIFT_ROWS(aes_state);
+    DLOG("vaesdf: @is_row state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vv.h:
```diff
@@ -0,0 +1,61 @@
+    DLOG("vaesdf: @is_row state(%" PRIxEG ")", PRVEG(aes_state));
+    // InvSubBytes - Apply S-box to every byte in the state
+    VAES_INV_SUB_BYTES(aes_state);
+    DLOG("vaesdf: @is_box state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vv.h:
```diff
@@ -0,0 +1,61 @@
+    VAES_INV_SUB_BYTES(aes_state);
+    DLOG("vaesdf: @is_box state(%" PRIxEG ")", PRVEG(aes_state));
+    // AddRoundKey (which is also InvAddRoundKey as it's xor)
+    DLOG("vaesdf: ik_sched round_key(%" PRIxEG ")", PRVEG(round_key));
```
Delete.

aswaterman commented on riscv/insns/vaesdf_vv.h:
```diff
@@ -0,0 +1,61 @@
+    // AddRoundKey (which is also InvAddRoundKey as it's xor)
+    DLOG("vaesdf: ik_sched round_key(%" PRIxEG ")", PRVEG(round_key));
+    EGU8x16_XOREQ(aes_state, round_key);
+    DLOG("vaesdf: @ik_add/@ioutput state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vs.h:
```diff
@@ -0,0 +1,69 @@
+#define PRVEG(X) PRVEGU8x16_LE(X)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vs.h:
```diff
@@ -0,0 +1,69 @@
+  {
+    DLOG("vaesef_vs vd#(%" PRIuREG ") vs2#(%" PRIuREG ") "
+         " vstart_eg(%" PRIuREG ") vl_eg(%" PRIuREG ")",
+         vd_num, vs2_num, vstart_eg, vl_eg);
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vs.h:
```diff
@@ -0,0 +1,69 @@
+
+    DLOG("vaesdm: @start state(%" PRIxEG "), scalar_key(%" PRIxEG ")"
+         " idx_eg(%" PRIuREG ")",
+         PRVEG(aes_state), PRVEG(scalar_key), idx_eg);
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vs.h:
```diff
@@ -0,0 +1,69 @@
+
+    // InvShiftRows - Rotate each row bytes by 0, 1, 2, 3 positions.
+    VAES_INV_SHIFT_ROWS(aes_state);
+    DLOG("vaesdm: @is_row state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vs.h:
```diff
@@ -0,0 +1,69 @@
+    DLOG("vaesdm: @is_row state(%" PRIxEG ")", PRVEG(aes_state));
+    // InvSubBytes - Apply S-box to every byte in the state
+    VAES_INV_SUB_BYTES(aes_state);
+    DLOG("vaesdm: @is_box state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vs.h:
```diff
@@ -0,0 +1,69 @@
+    VAES_INV_SUB_BYTES(aes_state);
+    DLOG("vaesdm: @is_box state(%" PRIxEG ")", PRVEG(aes_state));
+    // AddRoundKey (which is also InvAddRoundKey as it's xor)
+    DLOG("vaesdm: ik_sched round_key(%" PRIxEG ")", PRVEG(scalar_key));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vs.h:
```diff
@@ -0,0 +1,69 @@
+    // AddRoundKey (which is also InvAddRoundKey as it's xor)
+    DLOG("vaesdm: ik_sched round_key(%" PRIxEG ")", PRVEG(scalar_key));
+    EGU8x16_XOREQ(aes_state, scalar_key);
+    DLOG("vaesdm: @ik_add cipher state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vs.h:
```diff
@@ -0,0 +1,69 @@
+    DLOG("vaesdm: @ik_add cipher state(%" PRIxEG ")", PRVEG(aes_state));
+    // InvMixColumns
+    VAES_INV_MIX_COLUMNS(aes_state);
+    DLOG("vaesdm: @ioutput state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vv.h:
```diff
@@ -0,0 +1,63 @@
+#define PRVEG(X) PRVEGU8x16_LE(X)
+#else
+#define DLOG(...) (void)(0)
+#endif
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vv.h:
```diff
@@ -0,0 +1,63 @@
+  {
+    DLOG("vaesdm_vv vd#(%" PRIuREG ") vs2#(%" PRIuREG ")"
+         " vstart_eg(%" PRIuREG ") vl_eg(%" PRIuREG ")",
+         vd_num, vs2_num, vstart_eg, vl_eg);
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vv.h:
```diff
@@ -0,0 +1,63 @@
+
+    DLOG("vaesdm: @start state(%" PRIxEG "), key(%" PRIxEG ")"
+	 " idx_eg(%" PRIuREG ")",
+         PRVEG(aes_state), PRVEG(round_key), idx_eg);
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vv.h:
```diff
@@ -0,0 +1,63 @@
+
+    // InvShiftRows - Rotate each row bytes by 0, 1, 2, 3 positions.
+    VAES_INV_SHIFT_ROWS(aes_state);
+    DLOG("vaesdm: @is_row state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vv.h:
```diff
@@ -0,0 +1,63 @@
+    DLOG("vaesdm: @is_row state(%" PRIxEG ")", PRVEG(aes_state));
+    // InvSubBytes - Apply S-box to every byte in the state
+    VAES_INV_SUB_BYTES(aes_state);
+    DLOG("vaesdm: @is_box state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vv.h:
```diff
@@ -0,0 +1,63 @@
+    VAES_INV_SUB_BYTES(aes_state);
+    DLOG("vaesdm: @is_box state(%" PRIxEG ")", PRVEG(aes_state));
+    // AddRoundKey (which is also InvAddRoundKey as it's xor)
+    DLOG("vaesdm: ik_sched round_key(%" PRIxEG ")", PRVEG(round_key));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vv.h:
```diff
@@ -0,0 +1,63 @@
+    // AddRoundKey (which is also InvAddRoundKey as it's xor)
+    DLOG("vaesdm: ik_sched round_key(%" PRIxEG ")", PRVEG(round_key));
+    EGU8x16_XOREQ(aes_state, round_key);
+    DLOG("vaesdm: @ik_add cipher state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman commented on riscv/insns/vaesdm_vv.h:
```diff
@@ -0,0 +1,63 @@
+    DLOG("vaesdm: @ik_add cipher state(%" PRIxEG ")", PRVEG(aes_state));
+    // InvMixColumns
+    VAES_INV_MIX_COLUMNS(aes_state);
+    DLOG("vaesdm: @ioutput state(%" PRIxEG ")", PRVEG(aes_state));
```
Delete.

aswaterman replied:
OK, I'll just stop marking these, you get the idea.

egouriou-rivos commented:
I updated the commits, for a v5 version of this PR.

Semantically significant changes:
 - Added support for the new combo extensions "Zvknc" and "Zvksc", which are Zvkn/Zvks + Zvbc. They were added to the Zvk specification in version 0.9.7 on 2023-05-31 (https://github.com/riscv/riscv-crypto/releases/tag/v20230531). This is part of the "parse extensions" commit, in isa_parser.cc .
 - Followed Andrew's suggestion on how to handle the issue with element groups and WORDS_BIGENDIAN. The vector_unit::elt_group<> function invokes abort() if WORDS_BIGENDIAN is enabled and the isa_parser.cc logic emits an error if the extensions using element groups are enabled when WORDS_BIGENDIAN is in use.
 
Implementation changes:
 - Removed the debug logging support, per Andrew's request. See resolved comment above for more thoughts about this change.
 - Cleaned up the rotate logic following Andrew's suggestion, in particular introducing rotate_right/rotate_left templated functions in arith.h (part of the Zvk infrastructure commit).
 - Added prefixes (mostly "ZVK_") to some macros to reduce the risk of name collisions.
 - Moved some macros from zvk_ext_macros.h to the more specific extension header file.

As things stand, this commit series implements the current Zvk specification, version 0.9.7 of 2023-05-31, which is meant to be the freeze candidate.

egouriou-rivos commented:
Big thanks to Andrew for the suggestions and to the previous reviewers. More feedback is welcome.

aswaterman commented:
Thanks for addressing my last round of feedback. I’ll try to make another pass soon.

aswaterman reviewed (approved):
@egouriou-rivos I apologize for taking so long to provide a follow-up review.  I'm happy with this version.

aswaterman commented:
@egouriou-rivos Can you rebase on to the HEAD of master to resolve the conflicts?  You'll need to regenerate `encoding.h` yet again.  After you've done so and the CI checks are passing, I'll merge.  Thanks for your patience!

egouriou-rivos commented:
No worries about the delay, and thanks again for the reviews and suggestions.

I rebased on top of current /master. No changes to the commits, only resolving merge issues.

I regenerated riscv/encoding.h from riscv-opcodes /master (3ca60c5). Note that the previous encoding file was generated from a commit (8d70e77)  that I could not find in the riscv-opcodes upstream repository. It is not clear to me what else we are picking up between those two versions.

aswaterman commented:
Thanks!  There's a new batch of CI failures, but it looks like just a missing `#include`.

egouriou-rivos commented:
I uploaded a new patch series against the new /master (7e50839c) that addresses a couple of issues:
 - missing include in arith.h, linked to the use of std::is_unsigned<> from <type_traits>.
 - warning about signed/unsigned comparison in vcpop_v.h .

I missed both issues as I spend most of my time compiling macOS with Clang. The issues showed up when using GCC on Linux. Hopefully this will please the CI gods.

egouriou-rivos commented:
Checking those CI failures. I probably need to annotate some unused variables from the V macros, or derive new loop macros that do not declare those variables in the first place. I'll also rebase.

aswaterman commented:
Feel free to deploy the `UNUSED` macro liberally.

aswaterman commented:
Also, note I just merged #1364, which bumped `encoding.h` to the same version you've just bumped it to.

egouriou-rivos commented:
I uploaded a new version. It took me a little while to reproduce the failures.

Changes since last version:
- rebased on /master at 8b10de64, removing the need for commit updating the encodings;
 - Added a new VI_V_ULOOP macro in `v_ext_macros.h` for instructions of the form `<op> vd, vs2`. Adopted it for all Zvbb ".v" instructions that used to use VI_VX_ULOOP without using the `rs1` argument;
 - Fixed `vaesem_vv.h` to use the `VI_ZVK_VD_VS2_...` macro instead of `VD_VS1_VS2` which triggered the unused variable warning/error. All other Zvkned instructions already used that macro;
 - Removed the declaration of the `vsew` variable in the Zvk macros. None of those were used;
 - Added UNUSED in a few places in `vsm3c_vi.h` (Zvksh) and the three `vsha2...` instructions. The unused variables are declared when invoking a macro extracting elements from an element group. I would rather mark them UNUSED than have to define partial-extraction macros.

It would have been easier to sprinkle more UNUSED, but I appreciated the warnings emitted in the CI build. They did not reveal significant issues but some sloppiness.

With that I was able to build cleanly on macOS+Clang and Linux+GCC when adding the same warnings/errors that are used in the CI.

aswaterman commented:
Thanks for working through the remaining issues, @egouriou-rivos.  Time to pull the trigger.

egouriou-rivos commented:
Woohoo! Thank you.
